packages/adapters-library/src/core/utils/getBlockByDate.ts

```diff
@@ -0,0 +1,37 @@
+import { Block, JsonRpcProvider } from 'ethers'
+
+/**
+ * Performs a binary search to find the closest block by date.
+ */
+export async function getBlockByDate(
+  provider: JsonRpcProvider,
+  targetDate: Date,
+): Promise<Block | null> {
+  const targetTimestamp = targetDate.getTime()
+
+  // Get the latest block number
+  const latestBlockNumber = await provider.getBlockNumber()
+
+  // Perform a binary search between block 0 and the latest block
+  let lowerBound = 0
+  let upperBound = latestBlockNumber
+
+  while (lowerBound <= upperBound) {
+    const middle = Math.floor((lowerBound + upperBound) / 2)
+    const middleBlock = await provider.getBlock(middle)
+
+    if (!middleBlock) throw new Error(`Block ${middle} not found`)
+
+    // Found the block with the exact timestamp
+    if (middleBlock.timestamp === targetTimestamp) return middleBlock
```

Comment:
Is this trying to find a block in mainnet with exactly the same timestamp as an L2 block?

What happens if no block in mainnet has the exact same timestamp? This looks like it would throw.

Reply:
> Is this trying to find a block in mainnet with exactly the same timestamp as an L2 block?

Not the same timestamp, there's close to zero chances that this happens. The function `getBlockByDate` finds the block with **closest** timestamp as passed timestamp, using binary search. 

If no exact match (most probable case), it returns the closest (line 36).

Reply:
If you doubt this works, check the unit test, it covers the "not exactly same timestamp" case.

Reply:
I don't mind renaming the function `getBlockWithDateClosestFrom` if that helps.

Reply:
All good, thanks for clarifying.